pkg/destroy/azure/azure.go

```diff
@@ -0,0 +1,225 @@
+	// collect all the records from the zoneName
+	allPrivateRecords := sets.NewString()
+	for recordPages, err := recordsClient.ListByDNSZone(ctx, zoneGroup, zoneName, to.Int32Ptr(100), ""); recordPages.NotDone(); err = recordPages.NextWithContext(ctx) {
+		if err != nil {
+			return err
+		}
+		for _, record := range recordPages.Values() {
+			if t := toRecordType(to.String(record.Type)); t == dns.SOA || t == dns.NS {
+				continue
+			}
+			allPrivateRecords.Insert(fmt.Sprintf("%s.%s", to.String(record.Name), zoneName))
+		}
+	}
+
+	sharedZones, err := getSharedDNSZone(ctx, dnsClient, zoneName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find shared zone for %s", zoneName)
+	}
+	for _, sharedZone := range sharedZones {
+		logger.Debugf("removing matching private records from %s", sharedZone.Name)
+		for recordPages, err := recordsClient.ListByDNSZone(ctx, sharedZone.Group, sharedZone.Name, to.Int32Ptr(100), ""); recordPages.NotDone(); err = recordPages.NextWithContext(ctx) {
+			if err != nil {
+				return err
+			}
+			for _, record := range recordPages.Values() {
+				if allPrivateRecords.Has(fmt.Sprintf("%s.%s", to.String(record.Name), sharedZone.Name)) {
+					resp, err := recordsClient.Delete(ctx, sharedZone.Group, sharedZone.Name, to.String(record.Name), toRecordType(to.String(record.Type)), "")
+					if err != nil {
+						if wasNotFound(resp.Response) {
+							logger.WithField("record", to.String(record.Name)).Debug("already deleted")
+							continue
+						}
+						return errors.Wrapf(err, "failed to delete record %s in zone %s", to.String(record.Name), sharedZone.Name)
+					}
+					logger.WithField("record", to.String(record.Name)).Info("deleted")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// getSharedDNSZone returns the all parent public dns zone for privZoneName in decreasing order of closeness.
+func getSharedDNSZone(ctx context.Context, client dns.ZonesClient, privZoneName string) ([]dnsZone, error) {
+	domain := privZoneName
+	parents := sets.NewString(domain)
+	for {
+		idx := strings.Index(domain, ".")
+		if idx == -1 {
+			break
+		}
+		if len(domain[idx+1:]) > 0 {
+			parents.Insert(domain[idx+1:])
+		}
+		domain = domain[idx+1:]
+	}
+
+	allPublicZones := []dnsZone{}
+	for zonesPage, err := client.List(ctx, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+		if err != nil {
+			return nil, err
+		}
+		for _, zone := range zonesPage.Values() {
+			if zone.ZoneType == dns.Public && parents.Has(to.String(zone.Name)) {
+				allPublicZones = append(allPublicZones, dnsZone{Name: to.String(zone.Name), ID: to.String(zone.ID), Group: groupFromID(to.String(zone.ID)), Public: true})
+				continue
+			}
+		}
+	}
+	sort.Slice(allPublicZones, func(i, j int) bool { return len(allPublicZones[i].Name) > len(allPublicZones[j].Name) })
+	return allPublicZones, nil
+}
+
+type dnsZone struct {
```

Comment:
Do we want our own type, or can we use [`dns.Zone`][1] directly?

[1]: https://godoc.org/github.com/Azure/azure-sdk-for-go/services/preview/dns/mgmt/2018-03-01-preview/dns#Zone

Reply:
[`dns.Zone`][https://godoc.org/github.com/Azure/azure-sdk-for-go/services/preview/dns/mgmt/2018-03-01-preview/dns#Zone] has pointers for fields, doesn't have the resource group (that needs to be inferred from ID).

This small spare struct stores the details in easy to use way.